File: LocalPods/DialogView/DialogView/Core/Dialog+SpecialColors.m
//
//  Dialog+SpecialColors.m
//  TCLPlus
//
//  Created by OwenChen on 2021/1/4.
//  Copyright © 2021 TCL Electronics Holdings Ltd. All rights reserved.
//

#import "DialogModelEditor.h"

#import "Dialog+SpecialColors.h"


@implementation Dialog (SpecialColors)

/// 显示普通标题和副标题（颜色自定义）
+ (DialogView *)showDialogWithTitle:(NSString *)title
                         titleColor:(NSString *)titleColor
                           subtitle:(NSString *)subtitle
                      subtitleColor:(NSString *)subtitleColor
                            buttons:(NSArray<NSString *> *)buttons
                       buttonColors:(NSArray<NSString *> *)buttonColors
                        resultBlock:(DialogResultBlock)resultBlock {
    DialogView *view = [self dialogWithTitle:title
                                  titleColor:titleColor
                                    subtitle:subtitle
                               subtitleColor:subtitleColor
                                     buttons:buttons
                                buttonColors:buttonColors
                                 resultBlock:resultBlock];

    [self addDialogView:view];

    return view;
}

/// 显示普通标题和副标题（颜色自定义，副标题为富文本）
+ (DialogView *)showDialogWithTitle:(NSString *)title
                         titleColor:(NSString *)titleColor
                 attributedSubtitle:(NSAttributedString *)attributedSubtitle
                            buttons:(NSArray<NSString *> *)buttons
                       buttonColors:(NSArray<NSString *> *)buttonColors
                        resultBlock:(DialogResultBlock)resultBlock {
    DialogView *view = [self dialogWithTitle:title
                                  titleColor:titleColor
                          attributedSubtitle:attributedSubtitle
                                     buttons:buttons
                                buttonColors:buttonColors
                                 resultBlock:resultBlock];

    [self addDialogView:view];

    return view;
}

/// 普通标题和副标题（颜色自定义）
+ (DialogView *)dialogWithTitle:(NSString *)title
                     titleColor:(NSString *)titleColor
                       subtitle:(NSString *)subtitle
                  subtitleColor:(NSString *)subtitleColor
                        buttons:(NSArray<NSString *> *)buttons
                   buttonColors:(NSArray<NSString *> *)buttonColors
                    resultBlock:(DialogResultBlock)resultBlock {
    return [self dialogWithTitle:title
                      titleColor:titleColor
                        subtitle:subtitle
                   subtitleColor:subtitleColor
              attributedSubtitle:nil
                         buttons:buttons
                    buttonColors:buttonColors
                     resultBlock:resultBlock];
}

/// 普通标题和副标题（颜色自定义，副标题为富文本）
+ (DialogView *)dialogWithTitle:(NSString *)title
                     titleColor:(NSString *)titleColor
             attributedSubtitle:(NSAttributedString *)attributedSubtitle
                        buttons:(NSArray<NSString *> *)buttons
                   buttonColors:(NSArray<NSString *> *)buttonColors
                    resultBlock:(DialogResultBlock)resultBlock {
    return [self dialogWithTitle:title
                      titleColor:titleColor
                        subtitle:nil
                   subtitleColor:nil
              attributedSubtitle:attributedSubtitle
                         buttons:buttons
                    buttonColors:buttonColors
                     resultBlock:resultBlock];
}

#pragma mark - Private Methods

+ (DialogView *)dialogWithTitle:(NSString *)title
                     titleColor:(NSString *)titleColor
                       subtitle:(NSString *)subtitle
                  subtitleColor:(NSString *)subtitleColor
             attributedSubtitle:(NSAttributedString *)attributedSubtitle
                        buttons:(NSArray<NSString *> *)buttons
                   buttonColors:(NSArray<NSString *> *)buttonColors
                    resultBlock:(DialogResultBlock)resultBlock {
    NSMutableArray *array = [NSMutableArray array];
    if (title.length > 0) {
        TextDialogModel *textModel = [DialogModelEditor createTextDialogModelWithTitle:title titleColor:titleColor];
        [array addObject:textModel];
    }

    if (subtitle.length > 0) {
        TextDialogModel *textModel1 = [DialogModelEditor createTextDialogModelWithSubtitle:subtitle subtitleColor:subtitleColor];
        [array addObject:textModel1];
    } else if (attributedSubtitle.string.length > 0) {
        TextDialogModel *textModel1 = [DialogModelEditor createTextDialogModelWithAttributedSubtitle:attributedSubtitle subtitleColor:nil];
        [array addObject:textModel1];
    }

    [DialogModelEditor createModelMarginWithModels:array existTitle:title.length > 0];

    if (buttons.count > 0) {
        ButtonsDialogModel *buttonsModel = [DialogModelEditor createButtonsDialogModelWithButtons:buttons buttonColors:buttonColors];
        [array addObject:buttonsModel];
    }

    DialogView *view = [[DialogView alloc] initWithDialogModels:array
                                                      animation:[Dialog sharedInstance].animationType
                                                       position:DialogPositionMiddle
                                                        sideTap:YES
                                                         bounce:YES];
    view.resultBlock = resultBlock;

    return view;
}

@end

File: LocalPods/DialogViewSwift/DialogViewSwift/Core/Dialog+SpecialColors.swift
//
//  Dialog+SpecialColors.swift
//  TCLHome
//
//  Created by huaizhang.chen on 2022/3/29.
//  Copyright © 2022 TCL Eagle Lab. All rights reserved.
//

import Foundation

struct DialogSpecialColor {
    var titleColor: String = "" // 标题颜色
    var subtitleColor: String = "" // 副标题颜色
    var buttonColors: [String] = [] // 按钮颜色
}

struct DialogSubtitle {
    var subtitle: String = "" // 副标题
    var attributedSubtitle: NSAttributedString? // 富文本形式的副标题
}

// MARK: 自动显示弹框

extension EagleLabKit where Base: Dialog {
    /// 显示普通标题和副标题（非富文本）（颜色自定义）
    /// - Parameters:
    ///   - title: 标题
    ///   - subtitle: 副标题（非富文本）
    ///   - buttons: 按钮
    ///   - specialColor: 文字颜色配置
    ///   - resultBlock: 结果回调
    /// - Returns: DialogView
    @discardableResult
    static func showDialog(title: String,
                           subtitle: String,
                           buttons: [String]?,
                           specialColor: DialogSpecialColor,
                           resultBlock: DialogView.DialogViewResult?) -> DialogView {
        let view = dialog(title: title,
                          subtitle: DialogSubtitle(subtitle: subtitle),
                          buttons: buttons,
                          specialColor: specialColor,
                          resultBlock: resultBlock)
        Dialog.addDialog(dialogView: view)
        return view
    }

    /// 显示普通标题和副标题（富文本）（颜色自定义）
    /// - Parameters:
    ///   - title: 标题
    ///   - attributedSubtitle: 副标题（富文本）
    ///   - buttons: 按钮
    ///   - specialColor: 文字颜色配置
    ///   - resultBlock: 结果回调
    /// - Returns: DialogView
    @discardableResult
    static func showDialog(title: String,
                           attributedSubtitle: NSAttributedString,
                           buttons: [String]?,
                           specialColor: DialogSpecialColor,
                           resultBlock: DialogView.DialogViewResult?) -> DialogView {
        let view = dialog(title: title,
                          subtitle: DialogSubtitle(attributedSubtitle: attributedSubtitle),
                          buttons: buttons,
                          specialColor: specialColor,
                          resultBlock: resultBlock)
        Dialog.addDialog(dialogView: view)
        return view
    }
}

// MARK: 生成弹框，不自动显示

extension EagleLabKit where Base: Dialog {
    /// 自定义各种文字颜色的弹框
    /// - Parameters:
    ///   - title: 标题
    ///   - subtitle: 副标题
    ///   - buttons: 按钮
    ///   - specialColor: 文字颜色配置
    ///   - resultBlock: 结果回调
    /// - Returns: DialogView
    static func dialog(title: String,
                       subtitle: DialogSubtitle,
                       buttons: [String]?,
                       specialColor: DialogSpecialColor,
                       resultBlock: DialogView.DialogViewResult?) -> DialogView {
        var array: [BasicDialogModel] = []

        if !title.isEmpty {
            let textModel = DialogModelEditor.createTextDialogModel(title: title,
                                                                    titleColor: specialColor.titleColor)
            array.append(textModel)
        }

        if !subtitle.subtitle.isEmpty {
            let textModel = DialogModelEditor.createTextDialogModel(subtitle: subtitle.subtitle,
                                                                    subtitleColor: specialColor.subtitleColor)
            array.append(textModel)
        } else if let attributedSubText = subtitle.attributedSubtitle, !attributedSubText.string.isEmpty {
            let textModel = DialogModelEditor.createTextDialogModel(attributedSubtitle: attributedSubText, subtitleColor: "")
            array.append(textModel)
        }

        DialogModelEditor.createModelMargin(models: &array, existTitle: !title.isEmpty)

        if let buttonsArray = buttons, !buttonsArray.isEmpty {
            let buttonsModel = DialogModelEditor.createButtonsDialogModel(buttons: buttonsArray,
                                                                          buttonColors: specialColor.buttonColors)
            array.append(buttonsModel)
        }

        let view = DialogView()
        view.configDialogView(dialogModels: array)
        view.dialogResult = resultBlock
        return view
    }
}
